src/common/boundary_condition_manager.hpp

```diff
@@ -0,0 +1,299 @@
+   * @param[in] ess_bdr_coef The essential BC value coefficient
+   * @param[in] state The finite element state to which the BC should be applied
+   * @param[in] component The component to set (-1 implies all components are set)
+   */
+  void addEssential(const std::set<int>& ess_bdr, serac::GeneralCoefficient ess_bdr_coef, FiniteElementState& state,
+                    const int component = -1);
+
+  /**
+   * @brief Set the natural boundary conditions from a list of boundary markers and a coefficient
+   *
+   * @param[in] nat_bdr The set of mesh attributes denoting a natural boundary
+   * @param[in] nat_bdr_coef The coefficient defining the natural boundary function
+   * @param[in] component The component to set (-1 implies all components are set)
+   */
+  void addNatural(const std::set<int>& nat_bdr, serac::GeneralCoefficient nat_bdr_coef, const int component = -1);
+
+  /**
+   * @brief Set a generic boundary condition from a list of boundary markers and a coefficient
+   *
+   * @tparam The type of the tag to use
+   * @param[in] bdr_attr The set of mesh attributes denoting a natural boundary
+   * @param[in] bdr_coef The coefficient defining the natural boundary function
+   * @param[in] tag The tag for the generic boundary condition, for identification purposes
+   * @param[in] component The component to set (-1 implies all components are set)
+   * @pre Template type "Tag" must be an enumeration
+   */
+  template <typename Tag>
+  void addGeneric(const std::set<int>& bdr_attr, serac::GeneralCoefficient bdr_coef, const Tag tag,
+                  const int component = -1)
+  {
+    other_bdr_.emplace_back(bdr_coef, component, bdr_attr, num_attrs_);
+    other_bdr_.back().setTag(tag);
+    all_dofs_valid_ = false;
+  }
+
+  /**
+   * @brief Set a list of true degrees of freedom from a coefficient
+   *
+   * @param[in] true_dofs The true degrees of freedom to set with a Dirichlet condition
+   * @param[in] ess_bdr_coef The coefficient that evaluates to the Dirichlet condition
+   * @param[in] component The component to set (-1 implies all components are set)
+   */
+  void addTrueDofs(const mfem::Array<int>& true_dofs, serac::GeneralCoefficient ess_bdr_coef, int component = -1);
+
+  /**
+   * @brief Returns all the degrees of freedom associated with all the essential BCs
+   * @return A const reference to the list of DOF indices, without duplicates and sorted
+   */
+  const mfem::Array<int>& allDofs() const
+  {
+    if (!all_dofs_valid_) {
+      updateAllDofs();
+    }
+    return all_dofs_;
+  }
+
+  /**
+   * @brief Eliminates all essential BCs from a matrix
+   * @param[inout] matrix The matrix to eliminate from, will be modified
+   * @return The eliminated matrix entries
+   * @note The sum of the eliminated matrix and the modified parameter is
+   * equal to the initial state of the parameter
+   */
+  std::unique_ptr<mfem::HypreParMatrix> eliminateAllFromMatrix(mfem::HypreParMatrix& matrix) const
```

Comment:
I would rename this `eliminateAllEssentialDofsFromMatrix`. I know it's verbose, but these calls are tricky enough I think it's worth it.